### ISSUE TITLE
Jobs updated by checkpoint are not getting run

### ISSUE THREAD

**Issue by jmeppley** [snakemake/snakemake#727]
**Snakemake version**
works:
 * 5.20.0
 * 5.25.0
 * 5.26.0
 * 5.26.1

fails:
 * HEAD (5.27.4+1.g4a63a269) 
 * 5.27.4
 * 5.27.0

**Describe the bug**
The test run for our [DTR-phage-pipeline](https://github.com/nanoporetech/DTR-phage-pipeline) has stopped working recently. It has a few checkpoints and 160+ total jobs, so it's not the simplest test case, but it did work in earlier version.

Instead of completing in one go, it stops after each checkpoint. DEbugging output indicates that the downstream jobs are getting updated, but somehow it stops anyway. Restarting snakemake continues where it left off.

**Logs**
Here is the end of the output from the first run:
```
checkpoint generate_bins_and_stats:
    input: test/test-output/25m/1D/v2/kmer_binning/bin_membership.tsv
    output: test/test-output/25m/1D/v2/kmer_binning/bin_stats.csv, test/test-output/25m/1D/v2/kmer_binning/bins
    jobid: 11
Downstream jobs will be updated after completion.

[Thu Nov  5 14:15:24 2020]
rule plot_umap_freq_map_bins:
    input: test/test-output/25m/1D/v2/kmer_binning/bin_membership.tsv
    output: test/test-output/25m/1D/v2/kmer_binning/kmer_comp.umap.bins.png
    jobid: 9
                                                                                                                                                                                                                 
python /mnt/data0/jmeppley/projects/DTR/TBD/DTR-phage-pipeline/scripts/plot_umap_with_hdbscan_labels.py -o test/test-output/25m/1D/v2/kmer_binning/kmer_comp.umap.bins.png -s 7 -a 0.2 test/test-output/25m/1D/v2/kmer_binning/bin_membership.tsv
No kaiju DB found in /media/sfodata/databases/kaiju/nr_euk.
Skipping Kaiju
No kaiju DB found in /media/sfodata/databases/kaiju/nr_euk.
Skipping Kaiju
Job counts:
        count   jobs
        1       generate_bins_and_stats
        1
Activating conda environment: /mnt/data0/jmeppley/projects/DTR/TBD/DTR-phage-pipeline/.snakemake/conda/53e51836
Updating job 12 (combine_bin_cluster_read_info).
Replace combine_bin_cluster_read_info with dynamic branch combine_bin_cluster_read_info
updating depending job all
Updating job 13 (create_bin_cluster_read_lists).
Replace create_bin_cluster_read_lists with dynamic branch create_bin_cluster_read_lists
updating depending job combine_all_polished_ref_reads
updating depending job aggregate_prodigal_statistics
updating depending job combine_bin_cluster_strand_counts_into_table
updating depending job combine_dtr_aligns
updating depending job all
Updating job 21 (build_bin_cluster_summary_table).
Replace build_bin_cluster_summary_table with dynamic branch build_bin_cluster_summary_table
updating depending job combine_all_draft_stats
Updating job 0 (all).
Replace all with dynamic branch all
[Thu Nov  5 14:15:25 2020]
Finished job 11.
11 of 86 steps (13%) done
[Thu Nov  5 14:15:26 2020]
Finished job 8.
12 of 86 steps (14%) done
Mapping 869 data points...
Saving test/test-output/25m/1D/v2/kmer_binning/kmer_comp.umap.bins.notext.png
Mapping 869 data points...
[Thu Nov  5 14:15:26 2020]
Finished job 7.
13 of 86 steps (15%) done
bin -1 (66 reads)
bin 0 (315 reads)
bin 1 (79 reads)
bin 2 (45 reads)
bin 3 (29 reads)
bin 4 (12 reads)
bin 5 (27 reads)
bin 6 (15 reads)
bin 7 (42 reads)
bin 8 (42 reads)
bin 9 (20 reads)
bin 10 (13 reads)
bin 11 (36 reads)
bin 12 (28 reads)
bin 13 (25 reads)
bin 14 (16 reads)
bin 15 (18 reads)
bin 16 (29 reads)
bin 17 (12 reads)
Saving test/test-output/25m/1D/v2/kmer_binning/kmer_comp.umap.bins.png
Saving test/test-output/25m/1D/v2/kmer_binning/kmer_comp.umap.bins.nolabels.png
[Thu Nov  5 14:15:31 2020]
Finished job 9.
14 of 86 steps (16%) done
Complete log: /mnt/data0/jmeppley/projects/DTR/TBD/DTR-phage-pipeline/.snakemake/log/2020-11-05T141437.844467.snakemake.log
unlocking
removing lock
removing lock
removed all locks
```
Despite reporting 16% done, it just stops.

**Minimal example**
I'm working on a simpler example. For now it should be easy to reproduce by cloning the linked repo:

```
$ git clone https://github.com/nanoporetech/DTR-phage-pipeline
$ cd DTR-phage-pipeline
$ mamba create -p ./env -f ./environment.yaml
$ conda activate ./env
$ snakemake --use-conda --conda-frontend mamba -j 20 -p
```

**Additional context**
Oddly, if the first few rules (before the checkpoint) have already been completed, then the jobs after the checkpoint get run as they should. The erroneous stopping only occurs when starting with no output files.

**Comment by jmeppley** [comment]
Here is the minimal(ish) snakefile for this bug:

Snakefile:
```
INPUT_FILE = 'Snakefile'
OUTPUT_ROOT = Path('checkpoint-test')

BINS_ROOT                   = OUTPUT_ROOT / 'bins'
BIN_DIR_TEMPLATE            = OUTPUT_ROOT / 'bins' / '{bin_id}'
BIN_FILE_TEMPLATE           = OUTPUT_ROOT / 'bins' / '{bin_id}' / 'read_list.txt'
BIN_FILE2_TEMPLATE          = OUTPUT_ROOT / 'bins2' / '{bin_id}' / '{bin_id}.clust.info.csv'

rule all:
    input: OUTPUT_ROOT / 'final_file'

rule rule1:
    input: INPUT_FILE
    output: OUTPUT_ROOT / 'file1'
    shell:
        'date > {output}'

checkpoint checkpoint1:
    " defines some bins "
    input: OUTPUT_ROOT / 'file1'
    output:
        bins_dir=directory(BINS_ROOT)
    run:
        # 20 bins
        for bin_id in range(1,20):
            # create bin specific subdir of BINS_ROOT
            bin_dir = str(BIN_DIR_TEMPLATE).format(bin_id=bin_id)
            os.makedirs(bin_dir, exist_ok=True)

            # create a bin specific output file
            rlist_fn = str(BIN_FILE_TEMPLATE).format(bin_id=bin_id)
            with open(rlist_fn, 'w') as fh:
                fh.write('{bin_id}\n')

def expand_template_from_bins(wildcards, template):
    # get dir through checkpoints to throw Exception if checkpoint is pending
    checkpoint_dir = checkpoints.checkpoint1.get(**wildcards).output
    # get bins from files
    bins, = glob_wildcards(BIN_FILE_TEMPLATE)
    # expand template
    return expand(str(template), bin_id=bins)

# the second input is key here
rule rule2:
    input:
        bin_file=BIN_FILE_TEMPLATE,
        other=OUTPUT_ROOT / 'file1'
    output: BIN_FILE2_TEMPLATE
    shell:
        'date > {output}'

rule final_rule:
    """ access all bin files """
    input: lambda w: expand_template_from_bins(w, BIN_FILE2_TEMPLATE)
    output: OUTPUT_ROOT / 'final_file'
    shell: 'date > {output}'
```

And here is the output:
```
(environment) $ snakemake -j 20 -p
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 20
Rules claiming more threads will be scaled down.
Job counts:
        count   jobs
        1       all
        1       checkpoint1
        1       final_rule
        1       rule1
        4

[Thu Nov  5 19:31:10 2020]
rule rule1:
    input: Snakefile
    output: checkpoint-test/file1
    jobid: 3

date > checkpoint-test/file1
[Thu Nov  5 19:31:10 2020]
Finished job 3.
1 of 4 steps (25%) done

[Thu Nov  5 19:31:10 2020]
checkpoint checkpoint1:
    input: checkpoint-test/file1
    output: checkpoint-test/bins
    jobid: 2
Downstream jobs will be updated after completion.

Job counts:
        count   jobs
        1       checkpoint1
        1
Updating job 1 (final_rule).
[Thu Nov  5 19:31:11 2020]
Finished job 2.
2 of 23 steps (9%) done
Complete log: /mnt/data0/jmeppley/projects/DTR/TBD/.snakemake/log/2020-11-05T193110.620420.snakemake.log
```

If you run snakemake again, the workflow completes.

**Comment by jmeppley** [comment]
The bug first appears in commit 2e2c63a83f2511e5d79885036be1b9f2bf58f4db (#710)

**Comment by epruesse** [comment]
I believe this is affecting me as well. I have checkpoint-using workflows which curiously end with success, but didn't actually complete. Some jobs just don't get run. Restarting the workflow a few times is required to get everything to actually be executed.

**Comment by ASLeonard** [comment]
I also have had similar issues on checkpointing between 5.26.1 and 5.27.4, except with the opposite conclusion. I have multiple instances of the same checkpoint running on a cluster but with different wildcards. When the first finishes, all of them seem to simultaneously "end" according to snakemake, throwing errors as the outputs are not finished in the other instances. If I then re-run snakemake, another instance will complete and the remaining instances will throw errors. Instead, running with `--keep-incomplete` ends with an error, but the final output is (eventually) as expected across every instance.

I've included an abridged log below.

```
File "miniconda3/envs/cobra/lib/python3.8/site-packages/snakemake/shell.py", line 202, in __new__
    del cls._processes[jobid]
KeyError: 142

...

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "miniconda3/envs/cobra/lib/python3.8/concurrent/futures/_base.py", line 328, in _invoke_callbacks
    callback(self)
  File "miniconda3/envs/cobra/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 572, in _callback
    error_callback(job)
  File "miniconda3/envs/cobra/lib/python3.8/site-packages/snakemake/scheduler.py", line 557, in _error
    self._handle_error(job)
  File "miniconda3/envs/cobra/lib/python3.8/site-packages/snakemake/scheduler.py", line 567, in _handle_error
    self.running.remove(job)
KeyError: split_checkpoint_rule
```

**Comment by epruesse** [comment]
I thought I could work around this by just restarting the workflow until done. But of course there are jobs still out on the cluster, some of them long running, which messes things up royally.

**Comment by epruesse** [comment]
I downgraded to 5.26.1 but am still seeing the issue.

**Comment by pb-cdunn** [comment]
We have the same problem. It began when snakemake was updated in PyPI. We have tried various work-arounds, unsuccessfuly. What works is "restarting". In other words, snakemake fails to run the task that depends on the output of the "checkpoint" task,  but on a restart it goes fine. So this is partially a race-condition.

My theory: When depending on a function of "wildcards" derived from a "checkpoint", the dependencies from that function are not reevaluated after the checkpoint.

Hopefully **snakemake** will not be updated in Bioconda anytime soon.

**Comment by pb-cdunn** [comment]
https://anaconda.org/bioconda/snakemake -- `5.27.4`

Shoot. The latest is there, so I think we are broken in **Bioconda** now.

**Comment by isovic** [comment]
This is a really important issue, it should be resolved as soon as possible.

**Comment by cgr71ii** [comment]
Have you checked out if the files are being generated anyway? It's happening the same issue to me, but even when snakemake logs that have ended, the output files of "all" rule have been generated. As you have said, when I execute again snakemake continues in the last rule where it ended (in my case is not stopping just before my checkpoints, but in rules that are executed after my 1st checkpoint, but does not reach the 2nd and last checkpoint of my snakefile).

**Comment by ASLeonard** [comment]
Unsurprisingly this is still an issue in 5.28.

Starting a new pipeline from scratch ([pbipa](https://github.com/PacificBiosciences/pbipa)) there are multiple phases of sharding/gathering jobs through a checkpoint command. After each gather phase, the pipeline ends with a similar message first reported by @jmeppley of `4 of 47 steps (9%) done`. Just rerunning snakemake will then correctly continue sharding/analysing/gathering and then cleanly ending despite more jobs to be rerun.

Not sure if the difference is starting from scratch or the lack of multiple wildcard instantiations of a checkpoint, but this behaviour is opposite to what I reported earlier of the output being correct but snakemake failing anyway post-checkpoint.

**Comment by nick-youngblut** [comment]
I recently added 2 checkpoint rules in my snakemake (5.28.0) pipeline (previously not tracking all files produced by the rules), and now I seem to be getting the same error as reported above:

```
Traceback (most recent call last):
  File "/ebio/abt3_projects/software/dev/miniconda3_dev/envs/snakemake/lib/python3.6/site-packages/snakemake/__init__.py", line 723, in snakemake
    keepincomplete=keep_incomplete,
  File "/ebio/abt3_projects/software/dev/miniconda3_dev/envs/snakemake/lib/python3.6/site-packages/snakemake/workflow.py", line 961, in execute
    success = scheduler.schedule()
  File "/ebio/abt3_projects/software/dev/miniconda3_dev/envs/snakemake/lib/python3.6/site-packages/snakemake/scheduler.py", line 406, in schedule
    else self.job_selector_ilp(needrun)
  File "/ebio/abt3_projects/software/dev/miniconda3_dev/envs/snakemake/lib/python3.6/site-packages/snakemake/scheduler.py", line 588, in job_selector_ilp
    + lpSum(
  File "/ebio/abt3_projects/software/dev/miniconda3_dev/envs/snakemake/lib/python3.6/site-packages/snakemake/scheduler.py", line 580, in <listcomp>
    * lpSum([job.priority * scheduled_jobs[job] for job in jobs])
  File "/ebio/abt3_projects/software/dev/miniconda3_dev/envs/snakemake/lib/python3.6/site-packages/snakemake/jobs.py", line 1024, in priority
    return self.dag.priority(self)
  File "/ebio/abt3_projects/software/dev/miniconda3_dev/envs/snakemake/lib/python3.6/site-packages/snakemake/dag.py", line 383, in priority
    return self._priority[job]
KeyError: raw_fastqc
```

The `KeyError` differs from run to run, and it's never any of the checkpoint rule names.

More troubling, when I try to restart the snakemake pipeline, snakemake always states that the original Illumina read files (output from first rule that copies reads to temp directory) are incomplete. For example:

```
IncompleteFilesException:
The files below seem to be incomplete. If you are sure that certain files are not incomplete, mark them as complete with

    snakemake --cleanup-metadata <filenames>

To re-generate the files rerun your command with the --rerun-incomplete flag.
Incomplete files:
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/2_T3/R1_sub4map.fq.gz
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/2_T3/R2_sub4map.fq.gz
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/36_T6/R1_sub4map.fq.gz
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/7_T6/R1_sub4map.fq.gz
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/7_T6/R2_sub4map.fq.gz
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/93_T3/R1_sub4map.fq.gz
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/93_T3/R2_sub4map.fq.gz
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/93_T7/R1_sub4map.fq.gz
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/97_T6/R1_sub4map.fq.gz
/ebio/abt3_scratch/nyoungblut/LLMGA_33004609329/97_T6/R2_sub4map.fq.gz
```

If I run `snakemake --cleanup-metadata` on the files, I still get the same message when I run snakemake again. These read files are temp files that have already been deleted, so I don't see why snakemake considers them incomplete. 

Sadly, if I use `--rerun-incomplete`, the pipeline starts from the beginning, since snakemake thinks that these starting read files are not present (but for some reason requires them again)

**Comment by lpla** [comment]
Still happening on 5.29.0. I am testing with [Bitextor](https://github.com/bitextor/bitextor/blob/master/snakemake/Snakefile).

**Comment by pb-cdunn** [comment]
I don't understand why **snakemake** hasn't been reverted.

In Bioconda, we use **snakemake-minimal**, which fortunately still uses an old version. Within my company, we are now pinning to an older version.

**Comment by epruesse** [comment]
@johanneskoester Could you have a look at this and decide what needs to be done?

**Comment by martinsikora** [comment]
i can confirm seeing the same issue still on 5.30.1. on the first run snakemake completes without running the checkpoint jobs, only after re-starting the pipeline it completes all.

**Comment by pb-cdunn** [comment]
This is well over a month old now. In my opinion, **snakemake** should be reverted to `5.26.1`. To be completely clear, **snakemake** was not very useful to us until the introduction of `checkpoint`, allowing a dynamic workflow graph. This is not a minor regression.

@johanneskoester, if there is no response soon, I think we will be forced to fork **snakemake** and include the fork in our software distributions, as more and more users are complaining about this. I don't mind bugs, and I'm not criticizing the bug itself. I only ask for some discussion of its importance and possible resolution.

**Comment by jmeppley** [comment]
Poking around a bit more, I find that the workflow works just fine as long as the second input to rule2 is not an output of an earlier rule. If that second input is gone, or is present at the start of the run, then the workflow completed as it should.

**Comment by pb-cdunn** [comment]
Aha. In other words, as long as your workflow graph is not _truly_ dynamic. That's a major regression.

**Comment by jmeppley** [comment]
@pb-cdunn , not quite. It can depend on the output of the checkpoint, just not any other rules. So it is dynamic, just not very flexible. 

I'm also not suggesting that's OK, by any measure, I'm just trying to isolate the bug.

**Comment by ASLeonard** [comment]
@jmeppley, I'm not 100% sure I understand what you've said. Are you saying checkpoints will work correctly if the aggregating step has no other rule-based input (only no files or pre-existing files)?

For simplicity, I decided to actually just try the official snakemake docs for [checkpoints](https://snakemake.readthedocs.io/en/stable/_modules/snakemake/checkpoints.html), and case alone shows broken behaviour in 5.27.4. The final output of copying and pasting the example is 
```
6 of 7 steps (86%) done
```
and then the program ends "successfully". In this case, the aggregating function has no other dependencies, but still has issues. I was surprised a case this simple would cause issues, as I would have guessed any example in the docs would be related to internal tests.

**Comment by jmeppley** [comment]
@ASLeonard 

Yes, in the workflow i posted above, if you drop the second input from rule 2, it runs to completion. I tested it again yesterday on the development HEAD. 

Meanwhile, I share your surrprise that the example from the docs fails.

**Comment by jmeppley** [comment]
Hey, @ASLeonard @pb-cdunn @martinsikora @lpla @isovic @epruesse :

Can you try out PR #801? It solves my problem, but I want to know if it works for all of you, too.

**Comment by epruesse** [comment]
I was wondering whether it was ready yet. Will check.

**Comment by jmeppley** [comment]
Thanks! It's not passing ALL the automated checks yet, but I think it is promising.

**Comment by johanneskoester** [comment]
Thank you folks! This is now fixed by @jmeppley merged PR #801.

**Comment by epruesse** [comment]
Hate to say it, but it's not working for me. "168 of 12391 steps (1%) done" and exit 0 ...

**Comment by epruesse** [comment]
Should have made a backup of .snakemake so I could go back more easily. Sigh.

**Comment by jmeppley** [comment]
Well, da*n. Can you post a workflow (preferably a smallish one) that
doesn't finish in one go?

**Comment by epruesse** [comment]
It's a rather complicated one. Not sure I can create a minimal example easily. Any chance this is due to the database conversion? Anything that might have been persisted that could have caused this?

For now, I'm just kicking it over and over. Latest iteration left a ton of jobs sitting on the cluster when it exited (exit 0...) after saying something about the ILP solver failing.

**Comment by ASLeonard** [comment]
It completes the doc example, as well as the PB-IPA workflow which has multiple checkpoint stages, so **5.30.2** appears to have solved the "stopping after checkpoint" problem. 

These workflows never had the `keyerror/self.running.remove(job)` for me, but I won't get the chance to test my original workflow which hit that issue just yet.

**Comment by epruesse** [comment]
Is there a way for me to debug the issue? Some code I can insert to dump the DAG or so? I'm still having the issue. The part I don't get is why it doesn't even wait for the running jobs on the cluster to finish.

**Comment by jmeppley** [comment]
@epruesse, 

It seems like there are more than one of these rare checkpoint bugs. We managed to fix one here, but you're clearly seeing a different one. 

I found another (#817) and submitted a PR (#818). I have no idea if it's the same problem you're seeing, but it's probably worth trying out. 

If that doesn't work, open a new issue with as many details as you can provide (particularly the output logs and the snakefile), even if you've already entered them in this thread. I'm not an official developer, but I'm interested in getting checkpoints to work, so feel free to tag me in it.

**Comment by epruesse** [comment]
I don't seem to have any (of those) issues any more after removing `.snakemake` and upgrading to 5.32.0.
